NativePromise should be chainable so long that their resolve/reject type are convertible

jyavenard opened the pull request:
#### f69deb1f63af3a56d308fbfa58a8065ac27ee3f4
<pre>
NativePromise should be chainable so long that their resolve/reject type are convertible
<a href="https://bugs.webkit.org/show_bug.cgi?id=264028">https://bugs.webkit.org/show_bug.cgi?id=264028</a>
<a href="https://rdar.apple.com/117782213">rdar://117782213</a>

Reviewed by Youenn Fablet.

Make chainTo usable with different type of NativePromise producer.

This is achieved by creating a producer for the target promise and chaining
it to the current promise.
As the target producer is only ever moved, we can guarantee that it won&apos;t be
accessed on different thread concurrently and as such remove the need to
dispatch a new task.
We add a new whenSettled method that will immediately invoke the callback
when the source promise is settled.

For template argument type deduction to work properly, the Producer class
had to be moved outside the NativePromise class.

Update SourceBufferPrivate to use new conversion functionality.

Fly-by: replace all static_cast&lt;Ref&lt;NativePromise&gt;&gt; in NativePromise.h
with a private method to improve code readability.

* Source/WTF/wtf/NativePromise.h:
(WTF::invokeAsync):
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::removeCodedFrames):
(WebCore::SourceBufferPrivate::append):
(WebCore::SourceBufferPrivate::resetParserState):
(WebCore::SourceBufferPrivate::memoryPressure):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::applyConstraints):
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::takePhoto): Simplify,
remove unnecessary use of Producer.
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::runInCurrentRunLoopUntilDone):
(TestWebKitAPI::TEST):
(TestWebKitAPI::makeMoveableFunction): Use std::invoke as it&apos;s more modern.

Canonical link: <a href="https://commits.webkit.org/270990@main">https://commits.webkit.org/270990@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5584545204e7ce0863ef8c211e7307b063188098

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27026 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5642 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28262 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29235 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24710 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27479 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7506 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3034 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24570 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27288 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4463 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23197 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3903 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3985 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24195 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29871 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/23550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24667 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24605 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30186 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26210 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3999 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2193 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28102 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5462 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33663 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6487 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4464 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7290 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4371 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -43,6 +43,7 @@
 #include <wtf/Logger.h>
```
In commit description: "Mke" -> "Make"; "different type" -> "different types" or "a different type".

jyavenard committed:
NativePromise should be chainable so long that their resolve/reject type are convertible
https://bugs.webkit.org/show_bug.cgi?id=264028
rdar://117782213

Reviewed by Youenn Fablet.

Make chainTo usable with different type of NativePromise producer.

This is achieved by creating a producer for the target promise and chaining
it to the current promise.
As the target producer is only ever moved, we can guarantee that it won't be
accessed on different thread concurrently and as such remove the need to
dispatch a new task.
We add a new whenSettled method that will immediately invoke the callback
when the source promise is settled.

For template argument type deduction to work properly, the Producer class
had to be moved outside the NativePromise class.

Update SourceBufferPrivate to use new conversion functionality.

Fly-by: replace all static_cast<Ref<NativePromise>> in NativePromise.h
with a private method to improve code readability.

* Source/WTF/wtf/NativePromise.h:
(WTF::invokeAsync):
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::removeCodedFrames):
(WebCore::SourceBufferPrivate::append):
(WebCore::SourceBufferPrivate::resetParserState):
(WebCore::SourceBufferPrivate::memoryPressure):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::applyConstraints):
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::takePhoto): Simplify,
remove unnecessary use of Producer.
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::runInCurrentRunLoopUntilDone):
(TestWebKitAPI::TEST):
(TestWebKitAPI::makeMoveableFunction): Use std::invoke as it's more modern.

Canonical link: https://commits.webkit.org/270990@main
